Add recipe and ARM64 support for dependencies

HellAholic opened the pull request:
This pull request introduces several updates and improvements across multiple Conan recipes, with a particular focus on better ARM64 Windows support, dependency upgrades, and the addition of a new version for the `mpdecimal` library. The most significant changes include the addition of the `mpdecimal/4.0.1` recipe (with full build/test support), enhancements for ARM64 Windows builds in multiple libraries, and updates to dependency versions for better compatibility and security.

**New Library Version and ARM64 Windows Support**

* Added a new recipe for `mpdecimal/4.0.1`, including source retrieval, build logic for MSVC and Unix-like systems, packaging, and a test package to verify the build. This version is required for ARM64 Windows support in Python. [[1]](diffhunk://#diff-92ced11ca9b1a41d9e398b674fb739776280b0a56db81d1bc3e98f5adb5b829dR1-R4) [[2]](diffhunk://#diff-3a591ae6fa159dd5491e22a0c2bef0962c5dfe18129f1ade4a4b4d7c95342ce0R2-R3) [[3]](diffhunk://#diff-91455def2feed0bb6767cb2c92e698f871c13086b694d9400b1ecf90e4604e0eR1-R154) [[4]](diffhunk://#diff-f264cdfcc6dae586879fda774df6355b0dea4c431e9c94fbe9fc4077d28a96a2R1-R7) [[5]](diffhunk://#diff-4501242d6278272dd06f4423dc7b99c04c12109b2bc8dc246878cd654d271176R1-R25) [[6]](diffhunk://#diff-57fe24543584211e7fa469da953890a59093f1d2cce48c1520f68e3a4c915706R1-R29)

**ARM64 Windows Build Improvements**

* Updated the `cpython` recipe to require `mpdecimal/4.0.1` for ARM64 Windows builds, and improved logic for selecting `mpdecimal` versions based on platform and Python version. Also bumped `libffi` to `3.4.8` and `tk` to `8.6.13`. [[1]](diffhunk://#diff-902e47c71c4141d3938080326286ed09c238d3fad55af5f87c1e946567608f15L114-R119) [[2]](diffhunk://#diff-902e47c71c4141d3938080326286ed09c238d3fad55af5f87c1e946567608f15L139-R143)
* Improved ARM64 Windows support in the `hwloc` recipe by adding CMake flags and patching source files to handle ARM64 architecture. [[1]](diffhunk://#diff-0cc8f8338379c58f2941778863dbd2b6a79832349e7fb3cd699b16f9330509a5R60-R63) [[2]](diffhunk://#diff-0cc8f8338379c58f2941778863dbd2b6a79832349e7fb3cd699b16f9330509a5R77-R83)
* Enhanced ARM64 cross-compilation support in the `libffi` recipe by adding explicit build/host targets for ARM64 Windows.

**Dependency Version Updates**

* Updated the `onetbb` recipe to require `hwloc/2.11.1` instead of `2.9.3`.
* Updated the `pagmo2` and `pagmo2/pre_2.11` recipes to require `nlopt/2.10.0` instead of `2.7.1`. [[1]](diffhunk://#diff-66cb998213c679908ad25d34ca1005bbe9388f09e94d4f30aa0791a1213b29afL66-R64) [[2]](diffhunk://#diff-e77bb7d1cf77200e3e38f7326cfc17b088c75fe6051b02359c87cdd4d1026ac4L45-R45)
* Updated the `fmt` recipe to add MSVC-specific flags for Unicode handling and `/utf-8` compilation. [[1]](diffhunk://#diff-25b99d5905a7e7eaa414c5ae3fe09e93a5b26839ccc8bca5bbd932f05d3b814bR8) [[2]](diffhunk://#diff-25b99d5905a7e7eaa414c5ae3fe09e93a5b26839ccc8bca5bbd932f05d3b814bR159-R166)

**General Improvements and Fixes**

* Ensured proper file encoding when reading solution files in the `cpython` recipe for project discovery.
* Minor recipe improvements, such as using `rm_safe` for option removal and updating required Conan version in `pagmo2`. [[1]](diffhunk://#diff-66cb998213c679908ad25d34ca1005bbe9388f09e94d4f30aa0791a1213b29afL1-R9) [[2]](diffhunk://#diff-66cb998213c679908ad25d34ca1005bbe9388f09e94d4f30aa0791a1213b29afL51-R49)

**New Source Definitions**

* Added sources for new versions of the `shapely` library (`2.0.6` and `2.1.2`) in its conandata file.

HellAholic committed:
Add Shapely recipe and ARM64 support for dependencies

Introduces a new recipe for the Shapely Python library with test package and configuration files. Updates CPython, fmt, hwloc, libffi, mpdecimal, onetbb, pagmo2, tcl, and tk recipes to improve ARM64 (Windows) support, bump dependency versions, and add cross-compilation flags and patches where needed.

wawanbreton commented on recipes/hwloc/all/conanfile.py:
```diff
@@ -69,6 +73,14 @@ def generate(self):
         if self.settings.os == "Windows":
+            # ARM64 workaround: patch source files for ARM64 support
+            if self.settings.arch == "armv8":
+                from conan.tools.files import replace_in_file
```
This can probably be moved to top of the file

wawanbreton commented on recipes/shapely/all/test_package/conanbuild.bat:
I think this file (and all the scripts in this folder) are auto-generated by conan and thus should not be under version control, or is there a specific need ?

wawanbreton commented on recipes/tcl/all/conanfile.py:
```diff
@@ -156,14 +156,26 @@ def _build_nmake(self, targets):
                 targets=" ".join(targets),
-            ))
+            )
+            self.output.info(f"Running: {cmd}")
```
Aren't commands already displayed by `self.run` ? 🤔

wawanbreton reviewed (commented):
Some minor remarks to be adressed, but looks overall good. I am not a big fan of having all those exceptions, but we probably don't have a better solution. However we have to acknowledge that this will require more maintenance work in the future, and make sure it is worth it.

HellAholic committed:
Remove redundant log output

HellAholic committed:
Fix file encoding and variable unpacking issues

Specify UTF-8 encoding when reading the solution file to avoid decoding errors. Correct variable unpacking in package version comparison to prevent unused variable warnings.

HellAholic committed:
Refactor import of replace_in_file in conanfile.py

Moved the import of replace_in_file from inside the method to the top of the file for consistency and clarity.

HellAholic committed:
Update recipe to Conan v2 and modernize build tools

Migrated the pagmo2 recipe to Conan v2, replacing deprecated imports and APIs with their Conan v2 equivalents. Updated build and packaging logic to use new tools such as CMakeToolchain, CMakeDeps, and modern file operations. Removed legacy generator and property settings related to Conan v1 compatibility.

HellAholic committed:
Add missing newline to conandata.yml

Appended a newline at the end of recipes/shapely/all/conandata.yml to conform to file formatting standards.

HellAholic committed:
Remove Windows batch scripts from test_package as they are autogenerated by conan

HellAholic committed:
Remove unused imports in test_package conanfile

Cleaned up the conanfile.py in the test_package by removing unused imports to improve code clarity.

HellAholic committed:
Add newline at end of config.yml

Appended a newline to the end of recipes/shapely/config.yml for proper file formatting.

HellAholic committed:
Add mpdecimal 4.0.1 recipe

Introduces the mpdecimal 4.0.1 Conan recipe, including source data, build logic for MSVC and Unix-like systems, packaging steps, and a test package. Updates config.yml to register the new version.

HellAholic committed:
Remove custom package_id method

Remove duplicate method. Already in place.

HellAholic committed:
Revert unintentional changes

windows arm uses mpdecimal 4.0.x and not the 2.5.x, so the recipe should not be changed.

HellAholic committed:
Merge branch 'master' into CURA-12814_II

HellAholic committed:
Revert unnecessary change

unused variable name -> revert as it might be used in later versions, if not needed specifically, don't modify the recipe
